x/qgb/abci_test.go

```diff
@@ -526,3 +520,68 @@ func TestDataCommitmentCreationCatchup(t *testing.T) {
 		assert.Equal(t, want[i].Nonce, dc.Nonce)
 	}
 }
+
+// TestPruning tests the pruning mechanism by:
+// 1. Generating a set of attestations
+// 2. Running the QGB EndBlocker
+// 3. Verifying that the expired attestations are pruned
+func TestPruning(t *testing.T) {
+	input, ctx := testutil.SetupFiveValChain(t)
+	qgbKeeper := *input.QgbKeeper
+	// set the data commitment window
+	window := uint64(101)
+	qgbKeeper.SetParams(ctx, types.Params{DataCommitmentWindow: window})
+	initialBlockTime := ctx.BlockTime()
+	// make the interval between blocks being one hour
+	ctx = testutil.ExecuteQGBHeightsWithTime(ctx, qgbKeeper, 1, 1345, time.Hour)
+
+	// check that we created a number of attestations
+	assert.Equal(t, uint64(14), qgbKeeper.GetLatestAttestationNonce(ctx))
+
+	// check that no pruning occurs if the no attestation is expired
+	for nonce := uint64(1); nonce <= qgbKeeper.GetLatestAttestationNonce(ctx); nonce++ {
+		_, found, err := qgbKeeper.GetAttestationByNonce(ctx, nonce)
+		assert.NoError(t, err)
+		assert.True(t, found)
+	}
+
+	// continue executing heights
+	ctx = testutil.ExecuteQGBHeightsWithTime(ctx, qgbKeeper, 1345, 5000, time.Hour)
+
+	earliestAttestationNonce := qgbKeeper.GetEarliestAvailableAttestationNonce(ctx)
+	assert.Equal(t, uint64(38), earliestAttestationNonce)
```

Comment:
should this test have failed after switching to three weeks?

Reply:
Yeeeeeeeeeeeees, and it is failing :D meaning it works haha I will fix now

Reply:
[6396271](https://github.com/celestiaorg/celestia-app/pull/1799/commits/6396271f61d2fe02385ddcbe05a9180fd322fbe3)